<!--
Copyright 2014 Google Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->
<!DOCTYPE html>
<html lang="en">
<head>
  <script src="//ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js" charset="utf-8"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/underscore.js/1.6.0/underscore-min.js" charset="utf-8"></script>
  <script src="//rawgit.com/googlegenomics/api-client-javascript/master/googlegenomics.jquery.js"></script>
</head>
<body>

<div id="results" style="margin: 20px;">
</div>


<script type="text/javascript">
// Setup the genomics plugin
var clientId = window.location.hash.substring(1);
if (!clientId) {
  alert('A client ID is required. Please append one to the url: ' +
      'http://localhost:8000#my-client-id');
} else {
  $('#results').text('Loading');
  $.initGenomics({clientId: clientId});
}

//
// This example gets the read bases for a sample at specific a position
//
var datasetId = 376902546192; // This is the 1000 Genomes dataset ID
var sample = 'NA12872';
var referenceName = '22';
var referencePosition = 51003836;

// 1. First find the readset ID for the sample
$.genomicsAjax('/readsets/search?fields=readsets(id)', {
  method: 'POST',
  data: JSON.stringify({datasetIds: [datasetId], name: sample}),
  success: function(json) {
    var readsets = json.readsets || [];
    if (readsets.length != 1) {
      throw 'Searching for ' + sample +
          ' didn\'t return the right number of readsets';
    }

    var readsetId = readsets[0].id;


    // 2. Once we have the readset ID,
    // lookup the reads at the position we are interested in
    var params = {
      readsetIds: [readsetId],
      sequenceName: referenceName,
      sequenceStart: referencePosition,
      sequenceEnd: referencePosition,
      maxResults: '1024'
    };
    var readFields = 'reads(position,originalBases,cigar)';
    $.genomicsAjax('/reads/search?fields=' + readFields, {
      method: 'POST',
      data: JSON.stringify(params),
      success: function(json) {
        var reads = json.reads || [];
        var bases = _.map(reads, function(read) {
          // Note: This is simplistic -
          // the cigar should be considered for real code
          return read.originalBases[referencePosition - read.position]
        });
        var results = $('#results').text(sample + ' bases on ' + referenceName +
            ' at ' + referencePosition + ' are');
        _.each(_.countBy(bases), function(count, base) {
          $("<div/>").text(base + ": " + count).appendTo(results);
        });
      }
    });
  }
});


//
// This example gets the variants for a sample at a specific position
//

// 1. First find the call set ID for the sample
$.genomicsAjax('/callsets/search?fields=callSets(id)', {
  method: 'POST',
  data: JSON.stringify({variantSetIds: [datasetId], name: sample}),
  success: function(json) {
    var callSets = json.callSets || [];
    if (callSets.length != 1) {
      throw 'Searching for ' + sample +
          ' didn\'t return the right number of call sets';
    }

    var callSetId = callSets[0].id;


    // 2. Once we have the call set ID,
    // lookup the variants that overlap the position we are interested in
    var params = {
      callSetIds: [callSetId],
      referenceName: referenceName,
      // Note: currently, variants are 0-based and reads are 1-based,
      // reads will move to 0-based coordinates in the next version of the API
      start: referencePosition - 1,
      end: referencePosition
    };
    var variantFields = 'variants(names,referenceBases,alternateBases,' +
        'calls(genotype))';
    $.genomicsAjax('/variants/search?fields=' + variantFields, {
      method: 'POST',
      data: JSON.stringify(params),
      success: function(json) {
        var variant = json.variants[0];
        var variantName = variant.names[0];

        var genotype = _.map(variant.calls[0].genotype, function(g) {
          return g == 0 ? variant.referenceBases
              : variant.alternateBases[g - 1];
        });

        $('<div>').text('the called genotype is ' + genotype +
            ' for ' + variantName).appendTo($('#results'));
      }
    });
  }
});

</script>
</body>
</html>
